Revo: Flexi-Io Uart Support

henn1001 opened the pull request:
Probably breaks everything and has still a lot of errors but will update it here. Just wanted put it out here and draw attention to it :) 
Mostly a copy & paste from the brain because it´s using a similar pinout.

I think a lot of people out there are buying more revos, now that they are available for cheap. So this should be quite welcome. 

Fixes #261

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/d-ronin/dronin/617)

<!-- Reviewable:end -->

mlyle commented:
jenkins, test this please

mlyle commented:
@tracernz would you be willing to review?

mlyle commented:
@henn1001 -- please update the copyright year (2015-2016) and add the additional paragraph.

mlyle commented:
It also looks like some changes were maybe missed:

```
/var/lib/jenkins/jobs/dron/workspace@4/Nodes/master/ground/gcs/src/plugins/boards_openpilot/revolution.cpp: In member function ‘virtual Core::IBoardType::InputType Revolution::getInputOnPort(int)’:
/var/lib/jenkins/jobs/dron/workspace@4/Nodes/master/ground/gcs/src/plugins/boards_openpilot/revolution.cpp:223:21: error: ‘struct HwRevolution::DataFields’ has no member named ‘RcvrPort’
     switch(settings.RcvrPort) {
```

That said, I'm super excited about this.  Having another UART will make revo much more useful.

henn1001 commented:
Yeah i did not compile it myself yet - need to update the toolchain :)

mlyle commented:
jenkins, test this please

tracernz commented on flight/targets/revolution/board-info/board_hw_defs.c:
```diff
@@ -938,17 +989,66 @@ static const struct pios_usart_cfg pios_usart_dsm_hsum_flexi_cfg = {
-#include <pios_i2c_priv.h>
+
+/*
+ * FLEXI-IO PORT
```
Rx port?

mlyle commented:
Maybe time to remove the "WIP WIP WIP" from it and we can get it tested and reviewed, with some luck.

henn1001 commented:
I still have to look into the GCS side if it needs additional changes because of the name change of the rcv port. But so far it compiles both FW and GCS.

tracernz commented:
https://github.com/d-ronin/dRonin/blob/next/ground/gcs/src/plugins/boards_openpilot/revolution.cpp#L144-L233

henn1001 commented:
OK - so a few things i´d like to discuss:
- Naming convention:
  Should the change the portname to RxPort or stay with RcvrPort ? The official name from Openpilot is Flexi-Io port for the whole connector.
- Confusion:
  https://github.com/d-ronin/dRonin/pull/617/files#diff-59807a25d405a03b75d0a14fe511b018R1925
- Flexi-Io Mapping
  I changed the PWM output numbers slightly and adapted it to the brain but i´m not sure if this is wanted.
  http://opwiki.readthedocs.org/en/latest/user_manual/revo/revo.html  (Pinout Flexi-IO options)
  vs 
  http://brainfpv.com/support/hardware-setup/
  We should probably stay with the old definitions but then the option "PPM+UART+Outputs" get´s a bit more complicated. In this PR it is changed to Pin 10 = Output 7 | Pin 5= Output 12/PPM etc.  so i can subtract ports easily like it is done for the brain.
  https://github.com/d-ronin/dRonin/pull/617/files#diff-59807a25d405a03b75d0a14fe511b018R1758  
  The Timer banks in https://github.com/d-ronin/dRonin/pull/617/files#diff-ff82b679db198ed5d098fe4c7037126eR58 will probably have to adapt accordingly?
- I need to look at the frsky fast pwm rssi option again - did not quiet understand it yet. 

Also sorry for the USART rearrangement - make the PR/diff look a bit dirty 
Aaaand i should look into how to use reviewable.io :(

mlyle commented:
jenkins, test this please

@tracernz, you're the defacto revo expert.  Thoughts?

henn1001 commented:
jenkins, test this please

tracernz commented:
> @tracernz, you're the defacto revo expert. Thoughts?

Haha :-P

It's a bit hard to read the diff so I've been putting it off :/. I'll look it over tomorrow and test. Might be nice to put a board config page together so this can be made a bit easier/clearer to configure in GCS too, I can do that later in another PR if needed.

mlyle commented:
Ick, now it needs rebase; because of the sbus change.

tracernz commented:
Looks alright. Untested. Not sure if PPM was working before since it has the wrong timer channel?

---

Reviewed 1 of 5 files at r1, 4 of 5 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 756 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB5zNJEQE3-1g78q09T:-KB5zNJEQE3-1g78q09U:-599945314) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L756)):_
Is this necessary since this config is RX mode only?

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 797 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB5zh-j0EhJKwQu72Xw:-KB5zh-kb2SnhhlF4PJa:-1685268007) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L797)):_
Again, same.

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 1508 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB60HHKAGMouu9A2ZSS:-KB60HHLpjL95sCmgMj8:653881344) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L1508)):_
It would make more sense to re-order these so they're in ascending order rather than descending order right? It will be very confusing otherwise.. But I'm guessing this is required order to support PPM + outputs?

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 1925 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB62wUP-86TiIZmMDYM:-KB62wUP-86TiIZmMDYN:-1053382197) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L1925)):_
Brain 3?? Should be TIM_Channel_1 on PB14 right?

---

_[ground/gcs/src/plugins/boards_openpilot/revolution.cpp, line 200 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB64F5KEpFexNxAlU7g:-KB64F5KEpFexNxAlU7h:-1383491843) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/ground/gcs/src/plugins/boards_openpilot/revolution.cpp#L200)):_
Hmmm, this whole function (with and without the diff) is going to cause the wiring diagram to be quite wrong if the user has set ports up themselves before running the wizard :/. It should really only be an inverse of setInputOnPort above since that is the wiring configuration the diagram assumes.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617#-:-KB64sC_7Gj92gvSPriT:1042597305)_

<!-- Sent from Reviewable.io -->

henn1001 commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 1925 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB62wUP-86TiIZmMDYM:-KB7MyTLoO0065aOdx94:416179106) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L1925)):_
Yes i thought so too. 
But: https://github.com/henn1001/dRonin/blob/next/flight/targets/brain/board-info/board_hw_defs.c#L1455
Both targets use PB14 which should be Tim12Chan1. I think someone reported in IRC that his PPM wasn´t working with BRAIN? Maybe it´s a bug? or i don´t quiet get the meaning of this seperate Tim_Channel for PPM

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617)_

<!-- Sent from Reviewable.io -->

Commit:
Revo: Flexi-Io Support

henn1001 commented:
Bah that rebase for the NI-Sbus is going to be nasty. Need to figure out what the heck´s even going on lol.

mlyle commented:
:( sorry

Commit:
Revo: Flexi-Io Support - Rebase fixes

henn1001 commented:
Review status: 1 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 756 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB5zNJEQE3-1g78q09T:-KBDALWu5MkAhL4tlveH:-294772305) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L756)):_
fixed by  403

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 797 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB5zh-j0EhJKwQu72Xw:-KBDAQ9ZoVPcp4NLNSLP:1791625631) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L797)):_
fixed by 403

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617)_

<!-- Sent from Reviewable.io -->

henn1001 commented:
Review status: 1 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

_[flight/targets/revolution/board-info/board_hw_defs.c, line 1508 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB60HHKAGMouu9A2ZSS:-KBDFRLQ72b7gn8bNWCx:717656904) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/flight/targets/revolution/board-info/board_hw_defs.c#L1508)):_
Yeah i agree. Would probably be better if we stick to the intended functions https://opwiki.readthedocs.org/en/latest/user_manual/revo/revo.html . I´ll ditch the PPM + Uart + Output config for now unless there´s another way to selectively pick channels.
Would be cool if it will support octo + ppm + uart though in the future.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617)_

<!-- Sent from Reviewable.io -->

Commit:
Revo: Flexi-Io Support - Change outputs

Commit:
Revo: Flexi-Io Support - GCS fix

henn1001 commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

_[ground/gcs/src/plugins/boards_openpilot/revolution.cpp, line 200 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB64F5KEpFexNxAlU7g:-KBDaHPixSRSbuSXuQpC:751804143) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/ground/gcs/src/plugins/boards_openpilot/revolution.cpp#L200)):_
Not sure if i have fixed it now - please review again.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617)_

<!-- Sent from Reviewable.io -->

tracernz commented:
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

_[ground/gcs/src/plugins/boards_openpilot/revolution.cpp, line 200 [r4]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KB64F5KEpFexNxAlU7g:-KBGsTAWCDIhz48jhHxr:-363056077) ([raw file](https://github.com/d-ronin/dronin/blob/f4fb3cc488d8a8e0646799b7f7e69ac2796cbda8/ground/gcs/src/plugins/boards_openpilot/revolution.cpp#L200)):_
If you put all the RX on mainport then the below function should ignore FlexiPort. It should always ignore any setting that isn't made by setInputOnPort.

---

_[ground/gcs/src/plugins/boards_openpilot/revolution.cpp, line 175 [r6]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KBGsK2RHnmsmQEhASTu:-KBGsK2RHnmsmQEhASTv:-1740197210) ([raw file](https://github.com/d-ronin/dronin/blob/557218ce3553236415f04ef91dc276d536e211c7/ground/gcs/src/plugins/boards_openpilot/revolution.cpp#L175)):_
Can we do all the serial-rx on mainport to be consistent? Is there any reason not to? I'll update the (currently broken at least for S.Bus) connection diagram in another PR if you change it.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617)_

<!-- Sent from Reviewable.io -->

henn1001 commented:
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

_[ground/gcs/src/plugins/boards_openpilot/revolution.cpp, line 175 [r6]](https://reviewable.io:443/reviews/d-ronin/dronin/617#-KBGsK2RHnmsmQEhASTu:-KBHjyVVUGqd1MlTOoKA:201711473) ([raw file](https://github.com/d-ronin/dronin/blob/557218ce3553236415f04ef91dc276d536e211c7/ground/gcs/src/plugins/boards_openpilot/revolution.cpp#L175)):_
I´m not sure. Brain does everything on main port but for revo it says the main port won´t work for DSM because of the inverter.  Is brain using a two way inverter?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/617)_

<!-- Sent from Reviewable.io -->

Commit:
Revo: Flexi-Io Support - GCS..again

tracernz commented:
You're right, DSM binding won't work with the inverter, sorry.

pug398 commented:
@henn1001 On the rxport, the ppm+frsky is different from ppm+uart and selecting one of the frsky function for uart? Just curious :)

mlyle commented:
ppm+frsky attempts to demodulate frsky-RSSI high speed PWM on one of the receiver pins (I forget which).  It's poorly documented.  It seems to be broken on newer x4r-sb's and only really like d4rii.

WZ9V commented:
mlyle - your assessment is correct.  I've seen several different discussions attempting to figure out what the heck type of signal FrSky is outputting on that RSSI pin.  As far as I can tell nobody has come up with a reliable method of reading RSSI off of the pin.

mlyle commented:
Confirmed UART and PPM functionality.

mlyle commented:
Read the code.  seems sane for now.  Might not be perfect but doesn't make sense to gate it when it at least improves things.

tracernz commented:
Should raise an issue that DSM binding is broken then.

mlyle commented:
@tracernz on other targets with inverters we silently refuse to bind DSM if specified on an inverter port.

e.g. sparky2 does the exact same thing
 https://github.com/d-ronin/dRonin/blob/next/flight/targets/sparky2/fw/pios_board.c#L605

quanton drives an input, I think (this one we should fix (!!)) 
https://github.com/d-ronin/dRonin/blob/next/flight/targets/quanton/board-info/board_hw_defs.c#L750

does the same apply to CC3D?
https://github.com/d-ronin/dRonin/blob/next/flight/targets/cc3d/fw/pios_board.c#L254

and brain?
https://github.com/d-ronin/dRonin/blob/next/flight/targets/brain/fw/pios_board.c#L434

Would you like to open an issue and propose behavior?  I think at the least we should probably factor the binding-safety-check into pios_hal.

tracernz commented:
@mlyle the last commit makes the setup wizard configure DSM on the inverter port due to my incorrect suggestion earlier. Was waiting on that to be fixed ;-).
